bootloader/src/old_machine_bootloader_linux.c

```diff
@@ -0,0 +1,312 @@
+    fprintf(stderr, "malloc() failed\n");
+    exit(EXIT_FAILURE);
+  }
+  s->ptr[0] = '\0';
+}
+
+size_t writefunc(void *ptr, size_t size, size_t nmemb, struct response *s)
+{
+  size_t new_len = s->len + size*nmemb;
+  s->ptr = realloc(s->ptr, new_len+1);
+  if (s->ptr == NULL) {
+    fprintf(stderr, "realloc() failed\n");
+    exit(EXIT_FAILURE);
+  }
+  memcpy(s->ptr+s->len, ptr, size*nmemb);
+  s->ptr[new_len] = '\0';
+  s->len = new_len;
+
+  return size*nmemb;
+}
+
+
+char* executeCommand(char* commandLine){
+    FILE *fp;
+    char* path = (char *) malloc(sizeof(char) * 300);
+
+    /* Open the command for reading. */
+    fp = popen(commandLine, "r");
+    if (fp == NULL) {
+    printf("Failed to run command\n" );
+    exit(1);
+    }
+    /* Read the output a line at a time - output it. */
+    fgets(path, 300, fp);
+
+    /* close */
+    pclose(fp);
+
+    return path;
+}
+
+void checkHostName(int hostname)
+{
+    if (hostname == -1)
+    {
+        perror("gethostname");
+        exit(1);
+    }
+}
+
+// Returns host information corresponding to host name
+void checkHostEntry(struct hostent * hostentry)
+{
+    if (hostentry == NULL)
+    {
+        perror("gethostbyname");
+        exit(1);
+    }
+}
+
+// Converts space-delimited IPv4 addresses
+// to dotted-decimal format
+void checkIPbuffer(char *IPbuffer)
+{
+    if (NULL == IPbuffer)
+    {
+        perror("inet_ntoa");
+        exit(1);
+    }
+}
+
+// Replaces a single occurrence of substring
+char* replaceSubstringOnce(char* str, char* to_be_replaced, char* replacement) {
+    size_t str_size = strlen(str);
+    size_t to_be_replaced_size = strlen(to_be_replaced);
+    size_t replacement_size = strlen(replacement);
+    size_t result_size = str_size - to_be_replaced_size + replacement_size;
+    char *result_string = (char*)malloc(sizeof(char) * (result_size));
+    bool substringReplaced = false;
+
+    for (int i = 0; i < (int)result_size; i++ ){
+        if(substringReplaced){
+            break;
+        }
+        result_string[i] = str[i];
+        if(str[i] == to_be_replaced[0] && replacement_size != 0){
+            bool should_replace = true;
+            // Check if started iterating over string that will be replaced
+            for (int j = i; j < (i + to_be_replaced_size); j++){
+                if(str[j] != to_be_replaced[j - i]) {
+                    should_replace = false;
+                }
+            }
+            // If string that needs to be replaced is found - replace it
+            if (should_replace) {
+                for (int j = i; j < (i + replacement_size); j++){
+                    result_string[j] = replacement[j - i];
+                }
+                substringReplaced = true;
+            }
+        }
+    }
+    result_string[result_size] = '\0';
+    if(substringReplaced){
+        return result_string;
+    } else {
+        return str;
+    }
+}
+
+char *concatenate(size_t size, char *array[size], const char *joint){
```

Comment:
Why are you not using stdlib functions for this?

Reply:
memcpy is pretty similar to strcpy. Why is it worth refactoring?